cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -0,0 +1,212 @@
+        .filter(p -> p.getTetheringStatus() != TetheringStatus.REJECTED)
+        .collect(Collectors.toList());
+    } catch (IOException e) {
+      LOG.warn("Failed to get peer information", e);
+      return connectionInterval;
+    }
+    for (PeerInfo peer : peers) {
+      try {
+        Preconditions.checkArgument(peer.getEndpoint() != null,
+                                    "Peer %s doesn't have an endpoint", peer.getName());
+        String uri = CONNECT_CONTROL_CHANNEL + instanceName;
+        String lastMessageId = lastMessageIds.get(peer);
+        if (lastMessageId !=  null) {
+          uri = uri + "?messageId=" + URLEncoder.encode(lastMessageId, "UTF-8");
+        }
+        HttpResponse resp = TetheringUtils.sendHttpRequest(HttpMethod.GET, new URI(peer.getEndpoint())
+          .resolve(uri));
+        switch (resp.getResponseCode()) {
+          case HttpURLConnection.HTTP_OK:
+            handleResponse(resp, peer);
+            break;
+          case HttpURLConnection.HTTP_NOT_FOUND:
+            handleNotFound(peer);
+            break;
+          case HttpURLConnection.HTTP_FORBIDDEN:
+            handleForbidden(peer);
+            break;
+          default:
+            LOG.error("Peer {} returned unexpected error code {} body {}",
+                      peer.getName(), resp.getResponseCode(),
+                      resp.getResponseBodyAsString(StandardCharsets.UTF_8));
+        }
+      } catch (Exception e) {
+        LOG.debug("Failed to create control channel to {}", peer, e);
+      }
+    }
+    return connectionInterval;
+  }
+
+  private void handleNotFound(PeerInfo peerInfo) throws IOException {
+    // Update last connection timestamp.
+    store.updatePeerTimestamp(peerInfo.getName());
+
+    TetheringConnectionRequest tetherRequest = new TetheringConnectionRequest(instanceName,
+                                                                              peerInfo.getMetadata()
+                                                                          .getNamespaceAllocations());
+    try {
+      HttpResponse response = TetheringUtils.sendHttpRequest(HttpMethod.POST,
+                                                             new URI(peerInfo.getEndpoint())
+                                                            .resolve(TetheringClientHandler.CREATE_TETHER),
+                                                             GSON.toJson(tetherRequest));
+      if (response.getResponseCode() != 200) {
+        LOG.error("Failed to initiate tether with peer {}, response body: {}, code: {}",
+                  peerInfo.getName(), response.getResponseBody(), response.getResponseCode());
+      }
+    } catch (URISyntaxException | IOException e) {
+      LOG.error("Failed to send tether request to peer {}, endpoint {}",
+                peerInfo.getName(), peerInfo.getEndpoint());
+    }
+  }
+
+  private void initializeMessageIds() {
+    lastMessageIds = new HashMap<>();
+    List<String> topics;
```

Comment:
Seems like this should be called `peers` instead of `topics.

Reply:
fixed